protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,405 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::client::RelayedConnection;
+use crate::v2::RequestId;
+use futures::channel::mpsc;
+use futures::channel::oneshot;
+use futures::future::{ready, BoxFuture, Future, FutureExt, Ready};
+use futures::sink::SinkExt;
+use futures::stream::{Stream, StreamExt};
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use libp2p_core::transport::{ListenerEvent, TransportError};
+use libp2p_core::{PeerId, Transport};
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use thiserror::Error;
+
+/// A [`Transport`] enabling client relay capabilities.
+///
+/// Needs to be combined with another protocol, e.g. via
+/// [`OrTransport`](libp2p_core::transport::choice::OrTransport).
+///
+/// Allows the local node to:
+///
+/// 1. Establish relayed connections by dialing `/p2p-circuit` addresses.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_core::transport::choice::OrTransport;
+///    # use libp2p_relay::v2::client;
+///    let actual_transport = MemoryTransport::default();
+///    let (relay_transport, behaviour) = client::Client::new_transport_and_behaviour(
+///        PeerId::random(),
+///    );
+///    let transport = OrTransport::new(relay_transport, actual_transport);
+///    # let relay_id = PeerId::random();
+///    # let destination_id = PeerId::random();
+///    let dst_addr_via_relay = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(relay_id.into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
+///        .with(Protocol::P2p(destination_id.into())); // Destination peer id.
+///    transport.dial(dst_addr_via_relay).unwrap();
+///    ```
+///
+/// 3. Listen for incoming relayed connections via specific relay.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_core::transport::choice::OrTransport;
+///    # use libp2p_relay::v2::client;
+///    let actual_transport = MemoryTransport::default();
+///    let (relay_transport, behaviour) = client::Client::new_transport_and_behaviour(
+///        PeerId::random(),
+///    );
+///    let transport = OrTransport::new(relay_transport, actual_transport);
+///    let relay_addr = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
```

Comment:
```suggestion
///    # let relay_id = PeerId::random();
///    # let local_peer_id = PeerId::random();
///    let actual_transport = MemoryTransport::default();
///    let (relay_transport, behaviour) = client::Client::new_transport_and_behaviour(
///       local_peer_id,
///    );
///    let transport = OrTransport::new(relay_transport, actual_transport);
///    let relay_addr = Multiaddr::empty()
///        .with(Protocol::Memory(40)) // Relay address.
///        .with(Protocol::P2p(relay_id.into())) // Relay peer id.
```
To be consistent with the example above.

Reply:
:pray: